src/Storages/MergeTree/GinIndexStore.h

```diff
@@ -0,0 +1,267 @@
+    static GinIndexPostingsListPtr deserialize(ReadBuffer &buffer);
+};
+
+using GinIndexPostingsBuilderPtr = std::shared_ptr<GinIndexPostingsBuilder>;
+
+/// Container for all term's Gin Index Postings List Builder
+using GinIndexPostingsBuilderContainer = std::unordered_map<std::string, GinIndexPostingsBuilderPtr>;
+
+/// Container for postings lists for each segment
+using SegmentedPostingsListContainer = std::unordered_map<UInt32, GinIndexPostingsListPtr>;
+
+/// Postings lists and terms built from query string
+using PostingsCache = std::unordered_map<std::string, SegmentedPostingsListContainer>;
+using PostingsCachePtr = std::shared_ptr<PostingsCache>;
+
+/// Term dictionary information, which contains:
+struct TermDictionary
+{
+    /// .gin_post file offset of this segment's postings lists
+    UInt64 postings_start_offset;
+
+    /// .gin_dict file offset of this segment's term dictionaries
+    UInt64 item_dict_start_offset;
+
+    /// Finite State Transducer, which can be viewed as a map of <term, offset>
+    FST offsets;
+};
+
+using TermDictionaryPtr = std::shared_ptr<TermDictionary>;
+using TermDictionaries = std::unordered_map<UInt32, TermDictionaryPtr>;
+
+/// Gin Index Store which has Gin Index meta data for the corresponding Data Part
+class GinIndexStore
+{
+public:
+    GinIndexStore(const String& name_)
+        : name(name_)
+    {
+    }
+    GinIndexStore(const String& name_, DataPartStoragePtr storage_, DataPartStorageBuilderPtr data_part_storage_builder_, UInt64 max_digestion_size_)
+        : name(name_),
+        storage(storage_),
+        data_part_storage_builder(data_part_storage_builder_),
+        max_digestion_size(max_digestion_size_)
+    {
+    }
+
+    bool load();
+
+    /// Check existence by checking the existence of file .gin_seg
+    bool exists() const;
+
+    UInt32 getNextIDRange(const String &file_name, size_t n);
+
+    UInt32 getNextRowIDRange(size_t n);
+
+    UInt32 getNextSegmentID();
+
+    UInt32 getSegmentNum();
+
+    using GinIndexStorePtr = std::shared_ptr<GinIndexStore>;
+
+    void SetStorage(DataPartStoragePtr storage_)
+    {
+        storage = storage_;
+    }
+
+    GinIndexPostingsBuilderContainer& getPostings() { return current_postings; }
+
+    /// Check if we need to write segment to Gin index files
+    bool needToWrite() const;
+
+    /// Accumulate the size of text data which has been digested
+    void addSize(UInt64 sz) { current_size += sz; }
+
+    UInt64 getCurrentSegmentID() { return current_segment.segment_id;}
+
+    /// Do last segment writing
+    void finalize();
+
+    /// method for writing segment data to Gin index files
+    void writeSegment();
+
+    String getName() const {return name;}
+
+private:
+    void init_file_streams();
+private:
+    friend class GinIndexStoreReader;
+
+    String name;
+    DataPartStoragePtr storage;
+    DataPartStorageBuilderPtr data_part_storage_builder;
+
+    std::mutex gin_index_store_mutex;
```

Comment:
Seems like it requires us quite some time to digest this, it will be in our priority list. But I hope this will not block the PR merge this time :-)

Reply:
Thanks again, will consider this in the future.